scipy/stats/_continuous_distns.py

```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            numsamples = int(np.prod(shp))
+
+            # `out` is the array to be returned.  It is filled in in the
+            # loop below.
+            out = np.empty(self._size)
+
+            it = np.nditer([p, b],
+                           flags=['multi_index'],
+                           op_flags=[['readonly'], ['readonly']])
+            while not it.finished:
+                # Convert the iterator's multi_index into an index into the
+                # `out` array where the call to _rvs_scalar() will be stored.
+                # Where bc is True, we use a full slice; otherwise we use the
+                # index value from it.multi_index.  len(it.multi_index) might
+                # be less than len(bc), and in that case we want to align these
+                # two sequences to the right, so the loop variable j runs from
+                # -len(self._size) to 0.  This doesn't cause an IndexError, as
+                # bc[j] will be True in those cases where it.multi_index[j]
+                # would cause an IndexError.
+                idx = tuple((it.multi_index[j] if not bc[j] else slice(None))
+                            for j in range(-len(self._size), 0))
+                out[idx] = self._rvs_scalar(it[0], it[1],
+                                            numsamples).reshape(shp)
+                it.iternext()
+
+        if self._size == ():
+            out = out[()]
+        return out
+
+    def _rvs_scalar(self, p, b, numsamples=None):
+        invert_res = False
+        sz, rndm = numsamples, self._random_state
+        if not(sz):
+            sz = 1
+        if p < 0:
+            # note: if X is geninvgauss(p, b), then 1/X is geninvgauss(-p, b)
+            p = -p
+            invert_res = True
+        m = self._mode(p, b)
+        if (p >= 1) | (b > 1):
+            # ratio of uniforms with mode shift
+            a2 = -2 * (p + 1) / b - m
+            a1 = 2 * m * (p - 1) / b - 1
+            # find roots of x**3 + a2 * x**2 + a1 * x + m (Cardano's formula)
+            p1 = a1 - a2**2 / 3
+            q1 = 2 * a2**3 / 27 - a2 * a1 / 3 + m
+            phi = np.arccos(-q1 * np.sqrt(-27 / p1**3) / 2)
+            s1 = -np.sqrt(-4 * p1 / 3)
+            root1 = s1 * np.cos(phi / 3 + np.pi / 3) - a2 / 3
+            root2 = -s1 * np.cos(phi / 3) - a2 / 3
+            # root3 = s1 * np.cos(phi / 3 - np.pi / 3) - a2 / 3
+            vmin = (root1 - m) * np.sqrt(self.pdf(root1, p, b))
+            vmax = (root2 - m) * np.sqrt(self.pdf(root2, p, b))
+            umax = np.sqrt(self._pdf(m, p, b))
+            rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
+                                     umax=umax, vmin=vmin, vmax=vmax, c=m,
+                                     random_state=rndm)
+        elif b >= min(0.5, 2 * np.sqrt(1 - p) / 3):
+            # ratio of uniforms without mode shift
+            umax = np.sqrt(self._pdf(m, p, b))
+            s = sc.kv(p + 2, b) / sc.kv(p, b)
```

Comment:
You have 

            s = sc.kv(p + 2, b) / sc.kv(p, b)
            vmax = np.sqrt(self._pdf(self._mode(p + 2, b), p + 2, b) * s)

The version in the paper, translated to the code's variable names, would be

            xplus = ((1 + p) + sqrt((1 + p)**2 + b**2))/b
            vmax = xplus * np.sqrt(self._pdf(xplus, p, b))

After working through the algebra I see that mathematically they are the same, but I don't understand why you've expressed the calculation the way you did.  It looks like a confusing way to express the formula, and it also seems like it would be computationally more expensive--there are two additional calls to `kv` that the straight-forward version doesn't have.  Is there some advantage to your version that I don't see?

Reply:
no, i think it is better to use the version stated in the paper, fixed